src/CoreWCF.BuildTools/src/OperationParameterInjectionGenerator.Emitter.cs

```diff
@@ -0,0 +1,216 @@
+            {
+                foreach (var operationContractSpec in _generationSpec.OperationContractSpecs)
+                {
+                    EmitOperationContract(operationContractSpec);
+                }
+            }
+
+            private void EmitOperationContract(OperationContractSpec operationContractSpec)
+            {
+                string fileName = $"{operationContractSpec.ServiceContract.ContainingNamespace.ToDisplayString().Replace(".", "_")}_{operationContractSpec.ServiceContract.Name}_{operationContractSpec.MissingOperationContract.Name}.g.cs";
+                var dependencies = operationContractSpec.UserProvidedOperationContractImplementation.Parameters.Where(x => !operationContractSpec.MissingOperationContract.Parameters.Any(p =>
+                       p.IsMatchingParameter(x))).ToArray();
+
+                bool shouldGenerateAsyncAwait = SymbolEqualityComparer.Default.Equals(operationContractSpec.MissingOperationContract.ReturnType, _generationSpec.TaskSymbol)
+                    || (operationContractSpec.MissingOperationContract.ReturnType is INamedTypeSymbol symbol &&
+                    SymbolEqualityComparer.Default.Equals(symbol.ConstructedFrom, _generationSpec.GenericTaskSymbol));
+
+                Dictionary<ITypeSymbol, string> dependencyNames = new(SymbolEqualityComparer.Default);
+
+                string @async = shouldGenerateAsyncAwait
+                    ? "async "
+                    : string.Empty;
+
+                string @await = shouldGenerateAsyncAwait
+                    ? "await "
+                    : string.Empty;
+
+                string @return = (operationContractSpec.MissingOperationContract.ReturnsVoid || SymbolEqualityComparer.Default.Equals(operationContractSpec.MissingOperationContract.ReturnType, _generationSpec.TaskSymbol)) ?
+                    string.Empty
+                    : "return ";
+
+                string GetAccessibilityModifier(Accessibility accessibility) => accessibility switch
+                {
+                    Accessibility.Private => "private ",
+                    Accessibility.Protected => "protected ",
+                    Accessibility.Public => "public ",
+                    _ => "internal "
+                };
+
+                string returnType = operationContractSpec.MissingOperationContract.ReturnsVoid
+                    ? "void"
+                    : $"{operationContractSpec.MissingOperationContract.ReturnType}";
+
+                string parameters = string.Join(", ", operationContractSpec.MissingOperationContract.Parameters
+                    .Select(p => p.RefKind switch
+                    {
+                        RefKind.Ref => $"ref {p.Type} {p.Name}",
+                        RefKind.Out => $"out {p.Type} {p.Name}",
+                        _ => $"{p.Type} {p.Name}",
+                    }));
+
+                var indentor = new Indentor();
+                var builder = new StringBuilder();
+
```

Comment:
Is this method called concurrently for multiple instances? If not, then it might be worth caching this StringBuilder. So we're not reallocating every time. It has a Clear() method to reset it.

Reply:
Strictly in this PR code there is no concurrent access. However I think it could be called concurrently, it depends of the roslyn infrastructure integration within IDE I guess. I don't know if a mutex is set.
To be pragmatic I checked the Emitter implementation of STJ generator and nothing is done there to cache the StringBuilder. Please see there https://github.com/dotnet/runtime/blob/0f671237e9b148a371ffa41062a11a2e0f7b3e7d/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs#L149

Reply:
Okay, seems reasonable to leave it as there's ambiguity whether that's safe or not.